### ISSUE TITLE
[Windows] Ability to hot rebuild assets

### ISSUE THREAD

**Issue by IonutParau** [flutter/flutter#100445]
In my app, some important files are as assets and typically need to be modified for changes to occur.

With how Flutter currently is, I have to close and re-run the app to see those changes.

I would like something like hot restarting but for the asset system to be implemented, so assets could be recompiled while the app is running.

**Comment by maheshmnj** [comment]
Hi @IonutParau, Thanks for filing the issue. Can you please share on which target platform are you facing the issue?  

I verified on Android with the following steps and I see image updating with hot reload

1. created a new project 
2. added a image in `assets/image.png` 
3. Rendered image using `Image.asset()`
4. Start the app in debug mode
5. Added a new image at `assets/image2.png` 
6. updated the code sample to use image2 in place of image1
7. On Hot reload I see image 2 loading.

Please share the output of `flutter doctor -v`

<details>
<summary>flutter doctor -v (mac)</summary>

```
[✓] Flutter (Channel stable, 2.10.3, on macOS 12.1 21C52 darwin-arm, locale en-GB)
    • Flutter version 2.10.3 at /Users/mahesh/Documents/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 7e9793dee1 (20 hours ago), 2022-03-02 11:23:12 -0600
    • Engine revision bd539267b4
    • Dart version 2.16.1
    • DevTools version 2.9.2

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /Users/mahesh/Library/Android/sdk
    • Platform android-31, build-tools 31.0.0
    • ANDROID_HOME = /Users/mahesh/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 13.2.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • CocoaPods version 1.10.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)

[✓] IntelliJ IDEA Community Edition (version 2021.2.1)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 61.2.4
    • Dart plugin version 212.5080.8

[✓] VS Code (version 1.63.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.36.0

[✓] Connected device (3 available)
    • Redmi K20 Pro (mobile) • 192.168.1.5:5555 • android-arm64  • Android 11 (API 30)
    • macOS (desktop)        • macos            • darwin-arm64   • macOS 12.1 21C52 darwin-arm
    • Chrome (web)           • chrome           • web-javascript • Google Chrome 99.0.4844.51

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```

</details>

**Comment by maheshmnj** [comment]
Thanks for the info @IonutParau, Please share reproducible steps along with the output of `flutter doctor -v` to investigate this issue further.

**Comment by maheshmnj** [comment]
@IonutParau In order to investigate this issue further, we would need a minimal and complete code sample along with the reproducible steps.

**Comment by IonutParau** [comment]
Alright so, I'm using JSON files in the assets for translations currently, and the way I load them is like this:
```dart
List<File> get langs =>
    langDir.listSync().map<File>((item) => item as File).toList();
```

It gets re-fetched during function call so it is looking at how the assets are build (`langDir` is a Directory for the languages folder in the assets, I couldn't find a way to iterate them automatically using normal Asset Bundle)

When I modify the JSON files to correct them, and the Hot Restart, the app doesn't know of any modifications and when reading the files they are the exact same content. That is the reason I opened the issue, and make the request of a way to recompile the assets.

**Comment by maheshmnj** [comment]
> When I modify the JSON files to correct them, and then Hot Restart, the app doesn't know of any modifications and when reading the files they are the exact same content.

Since you are having an issue with the JSON file, I tried to reproduce the issue with the below code sample and I see that on hot reload or hot restart the latest file contents are returned.

<details>
<summary>code sample</summary>

##### assets/sample.json

```json
{
    "aps" : {
        "alert" : {
            "title" : "Hello world",
            "body" : "Bob wants to play poker",
            "action-loc-key" : "PLAY"
        },
        "badge" : 5
    },
    "acme1" : "bar",
    "acme2" : [ "bang",  "whiz" ]
}
```
**main.dart**
```dart
import 'dart:convert';

import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
        title: 'Flutter Demo',
        theme: ThemeData(
          primarySwatch: Colors.blue,
        ),
        home: const MyHomePage(
          title: 'Drawer demo',
        ));
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);
  final String title;
  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  Future<void> loadTitle() async {
    String data =
        await DefaultAssetBundle.of(context).loadString('assets/sample.json');
    Map jsonData = json.decode(data);
    setState(() {
      title = jsonData['aps']['alert']['title'];
    });
  }

  String title = '';

  @override
  void initState() {
    super.initState();
    loadTitle();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
        appBar: AppBar(
          title: Text(title),
        ),
        body: Center(
            child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            const Text('Tap the button to fetch the title from the json'),
            const SizedBox(
              height: 20,
            ),
            ElevatedButton(
                onPressed: loadTitle, child: const Text('fetch Title'))
          ],
        )));
  }
}

```
</details> 

<details>
<summary>flutter doctor -v (windows)</summary>

```
[√] Flutter (Channel stable, 2.10.3, on Microsoft Windows [Version 10.0.19044.1526], locale en-US)
    • Flutter version 2.10.3 at C:\flutter_sdk\stable
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 7e9793dee1 (14 hours ago), 2022-03-02 11:23:12 -0600
    • Engine revision bd539267b4
    • Dart version 2.16.1
    • DevTools version 2.9.2

[√] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at C:\Users\mahesh\AppData\Local\Android\sdk
    • Platform android-31, build-tools 31.0.0
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files\Google\Chrome\Application\chrome.exe

[√] Visual Studio - develop for Windows (Visual Studio Community 2019 16.11.4)
    • Visual Studio at C:\Program Files (x86)\Microsoft Visual Studio\2019\Community
    • Visual Studio Community 2019 version 16.11.31727.386
    • Windows 10 SDK version 10.0.19041.0

[√] Android Studio (version 2020.3)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)

[√] VS Code (version 1.64.2)
    • VS Code at C:\Users\mahesh\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.36.0

[√] Connected device (3 available)
    • Windows (desktop) • windows • windows-x64    • Microsoft Windows [Version 10.0.19044.1526]
    • Chrome (web)      • chrome  • web-javascript • Google Chrome 98.0.4758.102
    • Edge (web)        • edge    • web-javascript • Microsoft Edge 98.0.1108.56

[√] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```
<details>

**Comment by IonutParau** [comment]
Well this is awkward, I did try printing the contents of the JSON file that I fetched using the normal DefaultAssetBundle instead of File, and it printed it as normal. I then modified it and hot restarted and it worked perfectly.

This means the issue is with my application using the normal Dart IO file API.
I used it because I wanted it to automatically fetch all of them automatically, but I guess since it has this issue I should probably try a different method.

Thanks for the help!
